### PR TITLE
Move InterPodAffinity to skipped predicates for ecache temporarily

### LINKED ISSUE #67680: PodAffinity predicate should be invalidated for podAdd
**Is this a BUG REPORT or FEATURE REQUEST?**:
@kubernetes/sig-scheduling-bugs 

**What happened**:
In https://github.com/kubernetes/kubernetes/pull/67556, I am adding some new test cases with pod affinity and anti-affinity, and am seeing flaky test failures.

Debugging further, I see the following sequence of events from the test:
0. Pod A and Pod B have anti-affinity to each other using hostname topologyKey.
1. Pod A goes through the predicates successfully, fails during the volume assume (before pod assume)
2. Pod B goes through the predicates successfully, goes through pod assume and successfully binds to nodeA.
3. Pod A starts going through the scheduler again, it skips the PodAffinity predicate and uses the ecache value, which is true for all nodes.
4. Pod A gets bound to nodeA, violating the pod anti-affinity policy.

**What you expected to happen**:
PodAffinity predicate should be invalidated after Pod B is assumed.

**How to reproduce it (as minimally and precisely as possible)**:
From the PR, run repeatedly until failure:
```
make test-integration WHAT=./test/integration/scheduler GOFLAGS="-v" KUBE_TEST_ARGS="-race -run ^TestVolumeBindingWith"
```
It takes ~20 iterations to flake.  I added extra log statements in the PodAffinity predicate to determine that the predicate was not being called on the second run.

Edit: I updated the PR to disable the ecache for the test.  To reproduce, you'll have to modify the test to enable ecache again.

### PR DESCRIPTION
**Pull request by resouer**
**What this PR does / why we need it**:

The original assumption about AntiAffinity invalidation does not stand, see: https://github.com/kubernetes/kubernetes/issues/67680#issuecomment-415471575

Thus we need to add back the invalidate operation for it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67680

**Special notes for your reviewer**:
cc @msau42 

**Release note**:

```release-note
Invalidate anti-affinity on pod add
```

### CODE REVIEW

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (msau42) [inline pkg/scheduler/core/equivalence/eqivalence.go]: Is there a similar issue for PodAffinity too?  The predicate on all other nodes except this one would have to be invalidated?

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Ecache data structure is like: ecache[node][pred][equivhash], for all the other nodes, they do not have cached info for this pod.

**Review by yastij — APPROVED** [review]
/lgtm

this mostly lgtm, @bsalamat can we make this land into 1.12 as we're near code freeze.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (k82cn) [inline pkg/scheduler/core/equivalence/eqivalence.go]: s/MAtachInterPodAffinity/MatchInterPodAffinity

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (k82cn) [inline pkg/scheduler/core/equivalence/eqivalence.go]: Pod Anti Affinity is zone, not only node :(

Comment (k82cn) [reply pkg/scheduler/core/equivalence/eqivalence.go]: In my remember, we only support one zone, `failureDomain`, right now; let me check codes.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (k82cn) [inline pkg/scheduler/core/equivalence/eqivalence.go]: If we remove anti-affinity predicates when pod added, when will cache be hit?

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Sadly, no :( 

That's why there's a TODO: we will invalidate this predicate on pod add
1. for nodes in the same failure domain only
2. for pods with self anti-affinity only

But I would prefer to implement this after we merge eCache into predicates (discussed in #67241).

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Yeah you got it (it's aggressive though, see my explaintion below)

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: ack :)

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (misterikkit) [inline pkg/scheduler/core/equivalence/eqivalence.go]: failure domain != topology

I don't know if we can put a bound on the "blast radius" of affinity and anti-affinity rules.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (misterikkit) [inline pkg/scheduler/core/equivalence/eqivalence.go]: Should this be an OR instead of AND?

Comment (misterikkit) [inline pkg/scheduler/core/equivalence/eqivalence.go]: Any other pod could have an affinity or anti-affinity rule that matches this pod, even if there are no affinity rules on the new pod.

**Review by misterikkit — CHANGES_REQUESTED** [review]
> Is there a similar issue for PodAffinity too? The predicate on all other nodes except this one would have to be invalidated?

I can't reply to the earlier comment because the code has since changed. Anyway, it looks like you addressed @msau42's comment by invalidating the predicate across all nodes.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (k82cn) [reply pkg/scheduler/core/equivalence/eqivalence.go]: can we skip it?

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Instead of adding exception for a specific predicate, I would prefer to combine "cheapPredicates" logic here.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Emm, for a newly added pod without affinity rule defined, it either went thru real predicates, or reused cached predicates of an existing pod. It seems in both cases, the affinity/anti-affinity assumption of this node (and this failure domain) will not be broke? Any exception in your mind?

The main concern here is we have to invalidate predicates for all nodes. But fine-grained control is only possible when we can access `v1.Node` in ecache.

Thus, I recalled you've also mentioned in #67241, I would suggest let's combine the cheap predicates logic in #67241 to this PR, and add `MatchInterPodAffinity` into the skip list temporarily. So we can ship these two fixes in this cycle and continue merging eCache into predicates work at the same time. cc @k82cn.  

Sounds reasonable?

Comment (msau42) [reply pkg/scheduler/core/equivalence/eqivalence.go]: An example that demonstrates Jonathan's point could be something like:

1. Pod A has anti-affinity to Pods with label "foo"
2. Pod A ran through predicates once, but hit some error somewhere
3. Pod B comes along with label "foo" (but no pod.Spec.Affinity specified) and gets scheduled on node X
4. The anti-affinity predicate for Pod A should be invalidated on node X

If this was affinity instead, the the affinity predicate should be invalidated on all other nodes except X (in the optimized case)

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Thanks Michelle. This is actually a independent case I'm trying to figure out recently. i.e. When a Pod went thru predicates but did not occur on the node. We should invalidate all predicates (or specific predicates at least) for that Pod equiv class.

Right now the place I can think of doing this is after assume() returns error. Not sure if there's other "hit some error somewhere" case in your mind?

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: We already have one in #67241, that's why I propose to combine these two. Thus the predicates will be classified by the reason why they are skipped.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (k82cn) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Prefer to skip `MatchInterPodAffinity` for eCache instead of covering it case by case :(

> Any exception in your mind?

Affinity itself :), If no pod on any host, then all nodes are fits; after that, pod affinity only
schedule to the node with that pod.

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: > Affinity itself :), If no pod on any host, then all nodes are fits; after that, pod affinity only
schedule to the node with that pod.

eCache is keyed by node, I don't think "all nodes are fits" is possible ... mind explain a bit?

My proposal is temporarily skip this predicate and other cheap ones as well. Fine-grained control can be shipped after we achieved predicates pkg refactoring.

Comment (k82cn) [reply pkg/scheduler/core/equivalence/eqivalence.go]: > Affinity itself

If no pod on host_1 nor host_2, the predicates of both hosts are fit; after one pod dispatches to host_1, only host_1 should be fits. For eCache, if pod fit host_1 and host_2, there'll be two record (keyed by node); but it'll not refresh it.

> Not sure if there's other "hit some error somewhere" case in your mind?

Can we stop list cases here? IMO, Skip `MatchInterPodAffinity` is safer for this release.

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: > If no pod on host_1 nor host_2, the predicates of both hosts are fit; after one pod dispatches to host_1, only host_1 should be fits. For eCache, if pod fit host_1 and host_2, there'll be two record (keyed by node); but it'll not refresh it.

eCache is updated when pod is scheduled  to corresponding node. i.e. if `one pod dispatches to host_1`, then only `eCache[host_1]` is set, not sure why `there'll be two record (keyed by node)`?

> Can we stop list cases here?

Sorry, but the context of that part of discussion is about an exception affects ALL predicates, i.e. when sth wrong happened before bind(), we will have stale eCache. Also see this re-scheduling comment: https://github.com/kubernetes/kubernetes/issues/67680#issuecomment-415635647 , I have a patch for this fix.

> Skip MatchInterPodAffinity is safer for this release.

I can confirm this conclusion has been reached.

**Comment by resouer** [comment]
> I can't reply to the earlier comment because the code has since changed. Anyway, it looks like you addressed @msau42's comment by invalidating the predicate across all nodes.

Unfortunately, this is necessary as that predicate affects all nodes with same topology key. While we don't have node labels during predicate invalidation. Note that this can be fixed after we merged eCache into predicates. So I've left skipped it as well as cheap predicates as a all-in-one workaround for now.

File: pkg/scheduler/core/equivalence/eqivalence.go
Comment (msau42) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Is PodAdd also invoked when a pod is created? Not just when a pod is assumed?

Comment (k82cn) [reply pkg/scheduler/core/equivalence/eqivalence.go]: > eCache is updated when pod is scheduled to corresponding node. i.e. if one pod dispatches to host_1, then only eCache[host_1] is set

Sorry, did not check the detail of code. So only one node is ok for the eq pods? will scheduler dispatch all eq pods to the same node? or only the predicates on this node will not re-run?

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: Just re-use the predicates results of this node.

Comment (resouer) [reply pkg/scheduler/core/equivalence/eqivalence.go]: @msau42 Yes, but we have filter and separate handlers for assigned/unassigned Pods. The scheduler cache set & eCache invalidation are all handled by assigned Pod handler.

**Comment by resouer** [comment]
cc @k82cn, looks better now? It's now a combination of  basic logic of #67241 which align with same goal.

**Review by bsalamat — CHANGES_REQUESTED** [review]
Is this PR changed completely? The title of the PR has nothing to do with what it does.

**Comment by resouer** [comment]
@bsalamat Just changed the tile.

File: pkg/scheduler/factory/factory.go
Comment (k82cn) [inline pkg/scheduler/factory/factory.go]: Do we need this? As Pod affinity is skipped.

Comment (resouer) [reply pkg/scheduler/factory/factory.go]: Oh, should be rebase problem, just updated. Thanks!

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Please rename this to "eCacheBypassPredicates" to make it more clear why this list exists. You can then add "MatchInterPodAffinityPred" to the same set with a comment about the fact that it is temporarily skipped.

Comment (resouer) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Yes! Just updated.

**Comment by bsalamat** [comment]
@resouer do we still want to pursue this PR?

**Comment by resouer** [comment]
@bsalamat Though we are re-working eCache, the related issue still exists in master, so I think we still need to have this fix in if the idea looks generally good.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: s/ECacheMightHelpful/ECacheMightHelp/

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Do we have data that eCache helps with other predicates. I know that I measured performance of eCache for inter-pod affinity and I know that it helps there, but I have a feeling that this list is not backed by numbers. Please let me know if that is not true.

Comment (resouer) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Or do you think it make more sense to hold this change for now, as the on-going eCache refactoring is coming?

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (resouer) [reply pkg/scheduler/algorithm/predicates/predicates.go]: ack :-)

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates.go]: perfer to get the issue fixed; as someone said "Make It Work, Make It Right, Make It Fast" :)

Comment (resouer) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Sure, as we already have refactoring of eCache ready on dev branch. I think it's safe to hold this change, unless users complain.

**Comment by misterikkit** [comment]
Is this PR obsolete?

**Comment by resouer** [comment]
Close in favor of #71399

**Pull request closed without merging**
